### DISCUSSION TITLE
Lifetime of GPUBindGroupLayout and GPURender/ComputePipeline objects?

### DISCUSSION THREAD

**Discussion by juj** [gpuweb/gpuweb#1905] [Q&A] [answered]
1. If a GPUDevice is .destroy()ed, does that invalidate GPUBindGroupLayout objects created via `gpuDevice.createBindGroupLayout()`? In other words, are GPUBindGroupLayouts "parented" to the lifetime of the GPUDevice object they are created from?

2. Can GPUBindGroupLayouts from one GPUDevice instance be used with another GPUDevice instance?

3. It is expected that an application can create a lot of GPURenderPipeline and GPUComputePipeline objects (e.g. thousands). It is likely that those can consume a lot of CPU memory. There does not currently seem to exist any other way to delete those objects rather than to release references to them to let them GC. Should there exist an explicit `GPUPipelineBase.destroy()` function, or do they readily get GCd when a frame is over and all refs are dropped?
 
4. Do GPUBindGroupLayouts created with a call to `gpuRenderPipeline.getBindGroupLayout(idx)` have any association with the GPURenderPipeline they are created off of? In particular, if one acquires a GPUBindGroupLayout via `gpuRenderPipeline.getBindGroupLayout(idx)`, but then deletes the GPURenderPipeline that GPUBindGroupLayout object was created from (by dropping refs and letting it GC away), does deleting the GPURenderPipeline invalidate the GPUBindGroupLayout object? Or will that GPUBindGroupLayout object keep their "parent" GPURenderPipeline object alive until all GPUBindGroupLayout refs are also released?

**Comment by Kangz** [comment, marked as answer]
1. Yes in a sense. WebGPU only allows objects from the same device to interact together (otherwise a validation error is produced). So a `GPUBindGroupLayout` created on device D that's then destroyed will cause validation errors when used on object of that device (because the device is destroyed) and on objects of a different device (because it is a different device). The spec doesn't say anything about the content of objects of a destroyed device IIRC, but under the as-if rule, implementations are free to, and should, destroy as much as they can.
 2. No, because of the same-device rule.
 3. Right now they are only freed when the GC runs and destroys the corresponding `GPUPipeline` object. If that turns out to be an issue we could add a `.destroy()` method. However this has a cost because it will require validating the state of all pipelines used in a command buffer when it is submitted to the queue.
 4. The conceptually the `GPUPipelineBase` has a ref to the `GPUBindGroupLayout` but not the other way around so the pipeline can get GCed while the BGL is still reachable. However this might have limited use if / when we tweak the `getBindGroupLayout` behavior, see #1806 (we're trending towards something like solution 5)

**Reply by kainino0x** [reply]
4: Generally, GCs should never be observable. So, the GPUBindGroupLayout couldn't be invalidated by GC of the GPURenderPipeline. However we could end up adding a ref back from GPUBindGroupLayout to GPURenderPipeline if we need it internally (preventing the GPURenderPipeline from being destroyed).
